worker/suggestions.ts

```diff
@@ -0,0 +1,540 @@
+  for (const { name, description, star_count } of results) {
+    packages.set(name, { name, description, stars: star_count });
+  }
+}
+
+/** Fetch the meta data for a specific package/module and version from the S3
+ * bucket. */
+async function fetchMeta(pkg: string, ver: string): Promise<MetaJson> {
+  if (!packageMeta.has(pkg)) {
+    packageMeta.set(pkg, new Map());
+  }
+  const versionsMeta = packageMeta.get(pkg)!;
+  const res = await fetch(`${S3_BUCKET}${pkg}/versions/${ver}/meta/meta.json`);
+  if (res.status === 200) {
+    versionsMeta.set(ver, await res.json());
+  } else {
+    versionsMeta.set(ver, {
+      uploaded_at: "",
+      directory_listing: [],
+      upload_options: { type: "", repository: "", ref: "" },
+    });
+  }
+  return versionsMeta.get(ver)!;
+}
+
+/** Fetch the version data for a specific package/module from the S3 bucket. */
+async function fetchVersions(pkg: string): Promise<MetaVersionJson> {
+  const res = await fetch(`${S3_BUCKET}${pkg}/meta/versions.json`);
+  if (res.status === 200) {
+    versions.set(pkg, await res.json());
+  } else {
+    versions.set(pkg, { latest: "", versions: [] });
+  }
+  return versions.get(pkg)!;
+}
+
+/** Fetch package data from api.deno.land for a specific package/module. */
+async function fetchPackageData(pkg: string): Promise<PackageData> {
+  const res = await fetch(`https://api.deno.land/modules/${pkg}`);
+  if (res.status === 200) {
+    const { data: { name, description, star_count: stars } }:
+      ApiModuleResponse = await res.json();
+    packages.set(pkg, {
+      name,
+      description,
+      stars,
+    });
+  } else {
+    packages.set(pkg, {
+      name: pkg,
+      description: "",
+      stars: 0,
+    });
+  }
+  return packages.get(pkg)!;
+}
+
+/** Lazily update the module list for searching with fuse every 2 minutes and
+ * return a reference. */
+async function getFuse(): Promise<Fuse> {
+  const now = Date.now();
+  if (fuseUpdated + 120_000 < now) {
+    fuseUpdated = now;
+    const res = await fetch(MODULE_LIST);
+    if (res.status === 200) {
+      fuse = new Fuse(await res.json(), { includeScore: true, distance: 10 });
+    }
+  }
+  return fuse;
+}
+
+/** Git the initial package list to send to the client when no specific module
+ * is being searched for. */
+async function getInitialPackageList() {
+  if (!initialList) {
+    const res = await fetch(INITIAL_LIST);
+    if (res.status !== 200) {
+      throw new Error("bad api response");
+    }
+    const json: ApiModuleSearchResponse = await res.json();
+    cacheResponse(json);
+    initialList = toItems(json);
+  }
+  return initialList;
+}
+
+/** Given a package, return its latest version. */
+async function getLatestVersion(pkg: string): Promise<string> {
+  const data = versions.get(pkg) ?? await fetchVersions(pkg);
+  return data.latest;
+}
+
+function getMeta(pkg: string, ver: string): MetaJson | undefined {
+  return packageMeta.get(pkg)?.get(ver);
```

Comment:
Is this cache ever invalidated? If I understand right, newly published versions would not show up until the next time this script is restarted.

Reply:
No, but the `meta.json` is immutable, tied to a specific version, right?  Other things, like what versions there are for a package, are time limited cached, but a `meta.json` is considered immutable.  That is right, correct?

Reply:
Whoops, I commented this on the wrong function. I meant to comment on `fetchPackageData`. The `packages` cache is never invalidated as far as I can tell (or I can't find where it happens, in which case a comment would be nice).

Reply:
The package data isn't ever invalidated... I considered this a "small" risk, in that the only things that it would impact would be the package description and the star count.  I felt the natural lifecycle of a worker is sufficient.  I have made a note to that effect.

Reply:
Ah right, that totally makes sense. `versions` need to be invalidated though because they can change pretty frequently (latest version for example). Maybe just a 5m setInterval that wipes the cache?

Reply:
Yeah, good point about that...  I dislike `setInterval` though... :-) I can do like I do with the package cache and set a "wipe" timestamp and when a request is handled that requires the versions, wipe it before handling the request.